design_docs/analysis_task_template.md

```diff
@@ -0,0 +1,594 @@
+That is, it is the only file (other than `__init__.py`) in the base of the 
+`mpas_analysis` directory, making it easy to find.
+
+Note: we could put it in a subdirectory like `shared` or `template` but I think
+this makes sense.
+
+<h4> Design solution: Validation within Analysis Tasks <br>
+Date last modified: 2017/03/08 <br>
+Contributors: Xylar Asay-Davis
+</h4>
+The function `check` within each analysis task can be used to determine if
+necessary input files are present and/or if config options are set as expected.
+The template does not currently provide example of doing this but they could
+be added as comments.  My preference would be to make sure existing tasks
+provide examples of how this should be performed.
+
+Existing checks for missing observations files in `run_analysis.py` will be
+moved to individual analyses.  This will make clearer which checks correspond
+with which analysis tasks and will make clearer where such checks should be added
+within future analysis tasks.  Similarly, the addition of the `startDate` and
+`endDate` config options will be moved to the corresponding analysis tasks.
+
+<h4> Design solution: Analysis Continues even when Analysis Task Fails <br>
+Date last modified: 2017/03/08 <br>
+Contributors: Xylar Asay-Davis
+</h4>
+
+A try/except will be used around both `check` and `run` calls to make sure
+an error message and stack tracer are printed, but execution will continue
+for other tasks.
+
+<h4> Design solution: List of Tasks to Perform<br>
+Date last modified: 2017/03/08 <br>
+Contributors: Xylar Asay-Davis
+</h4>
+By having a standard set of implemented functions with standard names, it
+will be possible to import the analysis modules directly, e.g.
+`from mpas_analysis.ocean import ohc_timeseries`, instead of importing
+the function that runs the analysis, e.g.
+`from mpas_analysis.ocean.ohc_timeseries import ohc_timeseries`.  Then,
+each analysis task can be checked to see if it should be run based on
+the `generate` command-line or config option.  If so, its `check` 
+function will be run to make sure the configuration is right (and will
+print a warning if not).  If `check` passes, the analysis can be added
+to a list of functions to be run.  Later, a loop through the list
+can be used to run each analysis.
+
+Some analysis tasks require extra arguments (e.g. the field to be
+analyzed in the case of `ocean.modelvsobs` and the streams and variable
+maps for all analysis tasks).  These arguments will be stored as a
+dictionary of keyword arguments (kwargs) along side each 
+analysis-task module in the list of analyses to be run.
+
+
+<h3> Design and Implementation </h3>
+
+Implementation is in the branch: https://github.com/xylar/MPAS-Analysis/tree/analysis_task_template
+
+<h4> Implementation: Template for Analysis Tasks <br>
+Date last modified: 2017/03/08 <br>
+Contributors: Xylar Asay-Davis
+</h4>
+Here is the suggested template in full, intended to make discussion 
+of individual lines easier:
```

Comment:
Strange, something is weird with the renderer:  we don't get a nice code block when viewing the raw...